MaxText/tests/attention_test.py

```diff
@@ -1,110 +1,402 @@
+    b = self.global_batch_size
+    t = self.max_target_length
+    n = self.num_heads
+    d = self.head_dim
+    s = self.max_target_length
+    f = self.embed_dim
+    jnp_dtype = self.dtype
+
+    np.random.seed(0)
+    query_kernel = jnp.asarray(np.random.randn(f, n, d), dtype=jnp_dtype)
+    key_kernel = jnp.asarray(np.random.randn(f, n, d), dtype=jnp_dtype)
+    value_kernel = jnp.asarray(np.random.randn(f, n, d), dtype=jnp_dtype)
+    out_kernel = jnp.asarray(np.random.randn(n, d, f), dtype=jnp_dtype)
+
+    params = {
+        'query': {
+            'kernel': query_kernel
+        },
+        'key': {
+            'kernel': key_kernel
+        },
+        'value': {
+            'kernel': value_kernel
+        },
+        'out': {
+            'kernel': out_kernel
+        }
+    }
+    inputs_q = jnp.asarray(np.random.randn(b, t, f), dtype=jnp_dtype)
+    inputs_kv = jnp.asarray(np.random.randn(b, s, f), dtype=jnp_dtype)
+
+    mha_attention = MultiHeadDotProductAttention(
+        num_heads=self.num_heads,
+        head_dim=self.head_dim,
+        mesh=self.mesh,
+        dtype=jnp_dtype,
+        dropout_rate=self.cfg.dropout_rate,
+        name='self_attention',
     )
-    flash_output = self.attention.apply(
-            self.variable,
-            lnx,
-            lnx,
-            decoder_segment_ids = decoder_segment_ids,
-            attention_type='flash',
-            inputs_positions = decoder_positions,
-            mask = decoder_mask,
-            bias = None,
-            deterministic=True,
-            decode=False,
-            rngs={'aqt': self.rng}
+
+    _, _, _, decoder_positions = self.get_data(jnp_dtype)
+
+    out = mha_attention.apply(
+        {'params': freeze(params)},
+        inputs_q,
+        inputs_kv,
+        decoder_segment_ids=None,
+        inputs_positions=decoder_positions,
+        mask=None,
+        bias=None,
+        deterministic=True,
+        decode=False,
+        rngs={'aqt': self.rng})
+
+    query = jnp.einsum('btf,fnd->btnd', inputs_q, query_kernel)
+    query = LLaMARotaryEmbedding(
+        embedding_dims=self.head_dim,
+        name='query_rotary',
+    )(inputs=query, position=decoder_positions)
+
+    key = jnp.einsum('bsf,fnd->bsnd', inputs_kv, key_kernel)
+    key = LLaMARotaryEmbedding(
+        embedding_dims=self.head_dim,
+        name='key_rotary',
+    )(inputs=key, position=decoder_positions)
+    value = jnp.einsum('bsf,fnd->bsnd', inputs_kv, value_kernel)
+
+    logits = jnp.einsum('btnd,bsnd->bnts', query, key)
+    weights = jax.nn.softmax(logits, axis=-1)
+    combined = jnp.einsum('bnts,bsnd->btnd', weights, value)
+    expected = jnp.einsum('btnd,ndf->btf', combined, out_kernel)
+
+    out = np.asarray(out).astype(np.float16)
+    expected = np.asarray(expected).astype(np.float16)
+    np.testing.assert_allclose(out, expected, atol=1e-6)
+
+  def test_multiquery_attention(self):
+    """Test MultiQueryDotProductAttention."""
```

Comment:
Again I don't love the reimplementation. Shouldn't we test that MQA is the same as MHA (with queries replicated)?

Reply:
Stacking them into a method, is not readable.

Reply:
Refactoring this to equivalent test between GQA = MHA when num_kv_head = num_head